### ISSUE TITLE
Proposed API for Best-Effort Callback successor

### ISSUE THREAD

**Issue by danielballan** [bluesky/bluesky-mpl#15]
## The name

"Best-Effort Callback" is not very descriptive from a user point of view. Since this is, in practice, about  creating and managing matplotlib figures based on documents, let's just call it `Viewer`.

## Structure

The Viewer has multiple "outer tabs". Each outer tab can be used to (over-)plot one or more runs. The plots reside in one or more "inner tabs", contained by an outer tab.

## Proposed API

`Viewer` is a callable that can be used directly as a callback for consume streaming live data

```py
RE.subscribe(viewer)
```
or saved data.
```py
for name, doc in run.canonical(fill='yes'):
    viewer(name, doc)
```

A method that understands the API  of `run` is available as a terse way to display saved data.

```py
viewer.add_run(run)
```

By default all of the above target the currently-active outer tab. It is also possible to address a specific outer tab, to explicitly control what is plotted over what.

```py
viewer.add_tab(name)  # must be unique or ValueError is raised
```

A tab can be accessed by name
```py
viewer.tab[name].add_run(run)
```

or by position.
```py
viewer.tab[0].add_run(run)
```

A tab may be renamed.
```py
viewer.tab[name].rename(new_name)  # must be unique
```

**Comment by tacaswell** [comment]
This seems good to me.

An emplishment I would add would be for the outer tabs to keep a list of the uids it has plotted and an API for removing a run you no longer want to see.  This would also give a nice path to implement a FIFO plotting mode.

**Comment by prjemian** [comment]
Was thinking about this more.  Tom's suggestion to keep a list of uids fits
in with those thoughts.  The number of tabs could grow to be more than done
window managers can display well.  A scrolling tree view of tabs, with
collapsible subtree of uids would scale.

**Comment by danielballan** [comment]
That's a good idea. I'm also interested to see how users react to "overplot by default". That _seems_ to be what users generally want --- some UI to add a new tab when needed, otherwise overplot in the currently-active tab.

**Comment by danielballan** [comment]
@prjemian If you're up for taking this for an early spin, try out

```
git clone git@github.com:bluesky/bluesky-mpl
pip install -e .
ipython -i example_with_databroker.py
```

There many obvious omissions from the UI (closable tabs, etc.) but I'm interested in early reactions to the Python API.

**Comment by danielballan** [comment]
> An emplishment I would add would be for the outer tabs to keep a list of the uids it has plotted and an API for removing a run you no longer want to see.

Yes, and further, the low-level document-to-artist classes like our `Line`, `Grid`, and so on should get a `close()` method that removes their artists. Thus, `tab.remove(uid)` can cascade down into those classes which can clean themselves up.

**Comment by danielballan** [comment]
I just read through the documentation of [napari](https://napari.org/), and its API has much in common with the one proposed here. That kind of accidental "convergent evolution" is encouraging.

**Comment by jklynch** [comment]
Wow, what a bitchy comment I left. Here's version 2.

I would like to propose the name "RunViewer". I think it is helpful to say what will be viewed and while it's not perfect I think it is descriptive.

**Comment by danielballan** [comment]
Thanks, @jklynch. I agree that RunViewer is better and more clearly communicates the scope.

I could imagine the base RunViewer _eventually_ growing the ability to dispatch to things that aren't matplotlib-based. For example, maybe we want to use napari instead of `imshow` to view images, if it's available. This kind of thing might be a user-configurable extension. All that to say: avoiding "Figure" or "Matplotlib" in the name of this base object might turn out to be a forward-looking decision.
